src/mococrw/padding_mode.h

```diff
@@ -165,52 +133,50 @@ class MGF1 {
 };
 
 /**
- * @brief PSSPadding
+ * @brief OAEPPadding
  *
- * This class defines the parameters specific to the RSA PSS padding mode.
+ * This class defines the parameters specific to the OAEP padding mode:
  *
- * Defaults:
- * - Hash Function: SHA256
- * - Mask Generation Function: MGF1(<Hash Function>):
- *   By default, the specified hash function is also used for mask generation
- * - Salt length: Length of Hash Digest
+ * @warning: Because of the currently used implementation of OpenSSL (1.0.2), the label size should
+ *           be limited to maximum positive value of an integer (INT_MAX). This is a known bug that
+ *           was fixed in OpenSSL v1.1
  *
- * Currently only supports MGF1 objects for mask generation. If there ever will
- * be another mask generation function a second constructor may be added.
+ * Defaults:
+ *  - Hash Function: SHA256
+ *  - Mask Generation Function: MGF1(<Hash Function>)
+ *  - Label: Empty String
  */
-class PSSPadding: public RSASignaturePadding {
+class OAEPPadding : public RSAEncryptionPadding {
 public:
     /**
      * @brief Constructor
      * @param hashFunction Hash function to be used
-     * @param mgf1 MGF1 mask generation function to be used (if equals boost::none
-     *             by default MGF1(hashFunction) will be used)
-     * @param saltLength Length of salt to be used
+     * @param maskGenerationFunction mask generation function to be used (if equals nullptr
+     *                               by default MGF1(hashFunction) will be used)
+     * @param label Label to be used
      */
-    PSSPadding(openssl::DigestTypes hashFunction = openssl::DigestTypes::SHA256,
-               boost::optional<MGF1> maskGenerationFunction = boost::none,
-               boost::optional<int> saltLength = boost::none);
+    OAEPPadding(openssl::DigestTypes hashFunction = openssl::DigestTypes::SHA256,
+                std::shared_ptr<MaskGenerationFunction> maskGenerationFunction = nullptr,
+                const std::string& label="");
 
     /**
-     * @brief Copy Constructor
+     * @brief Copy Constrcutor
      */
-    PSSPadding(const PSSPadding& other);
+    OAEPPadding(const OAEPPadding& other);
 
     /**
      * @brief Copy Assignment
      */
-    PSSPadding& operator=(const PSSPadding& other);
+    OAEPPadding& operator=(const OAEPPadding& other);
 
     /**
      * @brief Destructor
      */
-    ~PSSPadding();
+    ~OAEPPadding();
```

Comment:
Well, I think the destructor is required because `Impl` is forward declared here. That is the destructor has to be declared once `Impl` is defined (in the source file). But you are right, probably it should be virtual...

Reply:
`override`?

Reply:
Virtual destructors don't work like virtual functions. afaik they are chained which means that the destructor of the base  as well as the destructor of the derived class are executed. That is, there is no override for destructors.